client/homebrew/utils/versionHistory.js

```diff
@@ -0,0 +1,120 @@
+// 	1: 2,			// 2 minutes
+// 	2: 10,			// 10 minutes
+// 	3: 60,			// 60 minutes
+// 	4: 12 * 60,     // 12 hours
+// 	5: 2 * 24 * 60  // 2 days
+// };
+// 
+// const GARBAGE_COLLECT_DELAY = 28 * 24 * 60; // 28 days
+
+
+// <=================== TEST VALUES STARTS ===================>
+
+// Test values
+const HISTORY_SAVE_DELAYS = {
+    0: 0,           // 0 minutes (if not specified)
+	1: 1,			// 1 minutes
+	2: 2,			// 2 minutes
+	3: 3,			// 3 minutes
+	4: 4,           // 4 minutes
+	5: 5            // 5 minutes
+};
+const GARBAGE_COLLECT_DELAY = 10; // 10 minutes
+
+// <==================== TEST VALUES ENDS ====================>
+
+
+
+const DEFAULT_STORED_BREW = {
+    shareId  : 'default_brew',
+    expireAt : '2000-01-01T00:00:00.000Z'
+};
+
+function getKeyBySlot(brew, slot){
+    return `${HISTORY_PREFIX}-${brew.shareId}-${slot}`;
+};
+
+function getVersionBySlot(brew, slot){
+    // Read stored brew data
+    // - If it exists, parse data to object
+    // - If it doesn't exist, pass default object
+    const key = getKeyBySlot(brew, slot);
+    const storedVersion = localStorage.getItem(key);
+    const output = storedVersion ? JSON.parse(storedVersion) : { ...DEFAULT_STORED_BREW, ...brew };
+    return output;
+};
+
+function updateStoredBrew(brew, slot=0){
+    const archiveBrew = {};
+
+    // Data from brew to be stored
+    archiveBrew.title = brew.title;
+    archiveBrew.text = brew.text;
+    archiveBrew.style = brew.style;
+    archiveBrew.version = brew.version;
+    archiveBrew.shareId = brew.shareId;
+
+    // Calculated values
+    archiveBrew.savedAt = new Date();
+    archiveBrew.expireAt = new Date();
+    archiveBrew.expireAt.setMinutes(archiveBrew.expireAt.getMinutes() + HISTORY_SAVE_DELAYS[slot]);
+    
+    // Store data
+    const key = getKeyBySlot(brew, slot);
+    localStorage.setItem(key, JSON.stringify(archiveBrew));
+    return;
+};
+
+export function updateHistory(brew) {
+    const numbers = [1,2,3,4,5];
+    const history = {};
+
+    // Load data from local storage to History object
+    numbers.forEach((i)=>{
+        history[i] = getVersionBySlot(brew, i);
+    });
+
+    // Walk each version position
+    numbers.toReversed().every((slot)=>{
+        const storedVersion = history[slot];
+
+        // If slot has expired, update all lower slots and break
+        if(new Date() >= new Date(storedVersion.expireAt)){
+            const slots = Array.from(Array(slot - 1).keys());
+            slots.toReversed().every((slot)=>{
+                const actualSlot = slot + 1;
+                // Move data from actualSlot to actualSlot + 1
+                updateStoredBrew({ ...history[actualSlot], shareId: brew.shareId }, actualSlot + 1);
+                // If first slot, fill with current data
+                if(actualSlot == 1) {
+                    updateStoredBrew(brew, 1);
+                    // Break after updating first slot
+                    return false;
+                }
```

Comment:
Thinking through what happens here if this is the very first entry.... Let me know if this is correct:

 `history` will be full of `DEFAULT_STORED_BREW`. The one in slot 5 will expire, then all the other `DEFAULT_STORED_BREW` in slots 1-4 will slide up, and the first entry will be in slot 1, set to expire in 1 minute.
 
 Next save, another `DEFAULT_STORED_BREW` expires, everything slides up, and a new entry in slot 1.
 
 So it will take 5 saves before any actual save data reaches slot 5. I think that's fine; shouldn't cause any user data to be expired prematurely.

Reply:
The way it's currently functioning, if the local storage key is empty, the current brew data is loaded instead with `expireAt` set in the past.
So the first time the function runs, it sees slot 5 is expired, overwrites it with the content of slot 4 - which is current brew values. But the act of writing it to slot 5 sets the `expireAt` to 5 days from current.
THEN the process of updating the lower slots does the same for slots 1 - 3, shuffling them down a slot (into 2 - 4) with their respective `expireAt` times - which, again, having been empty/no , are current brew values.
FINALLY, slot 1 is written with the current brew values.

Hopefully I've managed to explain it well enough: in short, the first time the function runs, we end up with the current data in every slot from 1 through 5, albeit with different values set for `expireAt`.